File: constraint-engine-ts/src/rules/nestjs-plugin-rules.ts
import { ConstraintRule, PluginValidationContext } from '../core/types';
import { createRule, ValidationHelpers } from '../core/rule-builder';

/**
 * NestJS Plugin Architecture Rules
 * Ported from Ruby with TypeScript improvements
 */

/**
 * Core plugin structure rules
 */
export const mustExtendBasePlugin = createRule()
  .id('must-extend-base-plugin')
  .name('Must Extend BaseActionPlugin')
  .description('All plugin classes must extend BaseActionPlugin')
  .severity('error')
  .category('architecture')
  .tags('plugins', 'inheritance')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    const classes = pluginContext.getClasses();
    
    const pluginClasses = classes.filter(cls => cls.endsWith('Plugin'));
    
    for (const pluginClass of pluginClasses) {
      if (!ValidationHelpers.contains(context.content, 'extends BaseActionPlugin')) {
        violations.push(ValidationHelpers.createViolation(
          'must-extend-base-plugin',
          'error',
          `Plugin class '${pluginClass}' must extend BaseActionPlugin`,
          pluginContext,
          { 
            suggestion: 'Add "extends BaseActionPlugin" to your plugin class declaration'
          }
        ));
      }
    }
    
    return violations;
  });

export const mustHaveInjectableDecorator = createRule()
  .id('must-have-injectable-decorator')
  .name('Must Have Injectable Decorator')
  .description('Plugin classes must have @Injectable() decorator')
  .severity('error')
  .category('architecture')
  .tags('plugins', 'decorators', 'nestjs')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    
    if (context.hasClass(/Plugin$/) && !context.hasDecorator('Injectable')) {
      violations.push(ValidationHelpers.createViolation(
        'must-have-injectable-decorator',
        'error',
        'Plugin class must have @Injectable() decorator',
        context,
        { 
          suggestion: 'Add @Injectable() decorator above your plugin class'
        }
      ));
    }
    
    return violations;
  });

export const mustHaveMetadata = createRule()
  .id('must-have-metadata')
  .name('Must Have Metadata Property')
  .description('Plugins must define readonly metadata property')
  .severity('error')
  .category('architecture')
  .tags('plugins', 'metadata')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    
    if (!ValidationHelpers.contains(context.content, 'readonly metadata: PluginMetadata')) {
      violations.push(ValidationHelpers.createViolation(
        'must-have-metadata',
        'error',
        'Plugin must have "readonly metadata: PluginMetadata" property',
        context,
        { 
          suggestion: 'Add: readonly metadata: PluginMetadata = { name: "...", version: "...", ... }'
        }
      ));
    }
    
    // Check required metadata fields
    const requiredFields = ['name', 'version', 'description', 'author', 'category'];
    const missingFields = requiredFields.filter(field => 
      !ValidationHelpers.contains(context.content, `${field}:`)
    );
    
    if (missingFields.length > 0) {
      violations.push(ValidationHelpers.createViolation(
        'must-have-metadata',
        'error',
        `Plugin metadata missing required fields: ${missingFields.join(', ')}`,
        context,
        { 
          suggestion: `Add missing fields to metadata: ${missingFields.join(', ')}`
        }
      ));
    }
    
    return violations;
  });

export const mustHaveSchema = createRule()
  .id('must-have-schema')
  .name('Must Have Schema Property')
  .description('Plugins must define readonly schema property')
  .severity('error')
  .category('architecture')
  .tags('plugins', 'schema')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    
    if (!ValidationHelpers.contains(context.content, 'readonly schema: ActionSchema')) {
      violations.push(ValidationHelpers.createViolation(
        'must-have-schema',
        'error',
        'Plugin must have "readonly schema: ActionSchema" property',
        context,
        { 
          suggestion: 'Add: readonly schema: ActionSchema = { request: RequestType, response: ResponseType }'
        }
      ));
    }
    
    return violations;
  });

export const mustHaveValidationRules = createRule()
  .id('must-have-validation-rules')
  .name('Must Have Validation Rules')
  .description('Plugins must define validation rules')
  .severity('error')
  .category('architecture')
  .tags('plugins', 'validation')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    
    if (!ValidationHelpers.contains(context.content, 'readonly validationRules: ValidationRule[]')) {
      violations.push(ValidationHelpers.createViolation(
        'must-have-validation-rules',
        'error',
        'Plugin must have "readonly validationRules: ValidationRule[]" property',
        context,
        { 
          suggestion: 'Add: readonly validationRules: ValidationRule[] = [...]'
        }
      ));
    }
    
    return violations;
  });

export const mustImplementExecuteMethod = createRule()
  .id('must-implement-execute')
  .name('Must Implement Execute Method')
  .description('Plugins must implement async execute method')
  .severity('error')
  .category('architecture')
  .tags('plugins', 'methods')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    
    if (context.hasClass(/Plugin$/) && !context.hasMethod(/^execute$/)) {
      violations.push(ValidationHelpers.createViolation(
        'must-implement-execute',
        'error',
        'Plugin must implement execute method',
        context,
        { 
          suggestion: 'Add: async execute(input: any): Promise<any> { ... }'
        }
      ));
    }
    
    // Check if execute method is async
    if (ValidationHelpers.contains(context.content, 'execute(') && 
        !ValidationHelpers.contains(context.content, 'async execute(')) {
      violations.push(ValidationHelpers.createViolation(
        'must-implement-execute',
        'error',
        'execute method must be async',
        context,
        { 
          suggestion: 'Add "async" keyword: async execute(input: any): Promise<any>'
        }
      ));
    }
    
    return violations;
  });

export const mustImplementHealthCheck = createRule()
  .id('must-implement-health-check')
  .name('Must Implement Health Check')
  .description('Plugins must implement async healthCheck method')
  .severity('error')
  .category('architecture')
  .tags('plugins', 'methods', 'monitoring')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    
    if (context.hasClass(/Plugin$/) && !context.hasMethod(/^healthCheck$/)) {
      violations.push(ValidationHelpers.createViolation(
        'must-implement-health-check',
        'error',
        'Plugin must implement healthCheck method',
        context,
        { 
          suggestion: 'Add: async healthCheck(): Promise<boolean> { return true; }'
        }
      ));
    }
    
    // Check return type
    if (ValidationHelpers.contains(context.content, 'healthCheck()') && 
        !ValidationHelpers.contains(context.content, 'Promise<boolean>')) {
      violations.push(ValidationHelpers.createViolation(
        'must-implement-health-check',
        'error',
        'healthCheck method must return Promise<boolean>',
        context,
        { 
          suggestion: 'Change signature to: async healthCheck(): Promise<boolean>'
        }
      ));
    }
    
    return violations;
  });

/**
 * Import and dependency rules
 */
export const mustImportRequiredModules = createRule()
  .id('must-import-required-modules')
  .name('Must Import Required Modules')
  .description('Plugins must import all required modules')
  .severity('error')
  .category('imports')
  .tags('plugins', 'imports')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    
    const requiredImports = [
      { module: '@nestjs/common', items: ['Injectable'] },
      { module: '../core/base-plugin.abstract', items: ['BaseActionPlugin'] },
      { module: '../core/plugin.interface', items: ['PluginMetadata', 'ActionSchema', 'ValidationRule'] }
    ];
    
    for (const required of requiredImports) {
      if (!context.hasImport(required.module)) {
        violations.push(ValidationHelpers.createViolation(
          'must-import-required-modules',
          'error',
          `Missing import from ${required.module}`,
          context,
          { 
            suggestion: `Add: import { ${required.items.join(', ')} } from '${required.module}';`
          }
        ));
      }
    }
    
    return violations;
  });

/**
 * Naming convention rules
 */
export const pluginFileNaming = createRule()
  .id('plugin-file-naming')
  .name('Plugin File Naming Convention')
  .description('Plugin files must follow naming convention')
  .severity('warning')
  .category('naming')
  .tags('plugins', 'conventions')
  .filePattern(/src\/plugins\/.*\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    
    if (!context.file.endsWith('.plugin.ts')) {
      violations.push(ValidationHelpers.createViolation(
        'plugin-file-naming',
        'warning',
        'Plugin files should end with .plugin.ts',
        context,
        { 
          suggestion: 'Rename file to follow pattern: feature-name.plugin.ts'
        }
      ));
    }
    
    return violations;
  });

export const pluginClassNaming = createRule()
  .id('plugin-class-naming')
  .name('Plugin Class Naming Convention')
  .description('Plugin classes must end with "Plugin"')
  .severity('error')
  .category('naming')
  .tags('plugins', 'conventions')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    const classes = pluginContext.getClasses();
    
    for (const className of classes) {
      if (!className.endsWith('Plugin')) {
        violations.push(ValidationHelpers.createViolation(
          'plugin-class-naming',
          'error',
          `Plugin class '${className}' must end with 'Plugin'`,
          context,
          { 
            suggestion: `Rename class to '${className}Plugin'`
          }
        ));
      }
    }
    
    return violations;
  });

export const metadataCategoryValidation = createRule()
  .id('metadata-category-validation')
  .name('Valid Metadata Category')
  .description('Plugin category must be valid')
  .severity('error')
  .category('validation')
  .tags('plugins', 'metadata')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    const validCategories = ['business', 'utility', 'integration', 'auth'];
    
    const categoryMatch = context.content.match(/category:\s*['"]([^'"]+)['"]/);
    if (categoryMatch) {
      const category = categoryMatch[1];
      if (!validCategories.includes(category)) {
        violations.push(ValidationHelpers.createViolation(
          'metadata-category-validation',
          'error',
          `Invalid category '${category}'. Must be one of: ${validCategories.join(', ')}`,
          context,
          { 
            suggestion: `Use valid category: ${validCategories.join(', ')}`
          }
        ));
      }
    }
    
    return violations;
  });

/**
 * Security rules
 */
export const noConsoleLogging = createRule()
  .id('no-console-logging')
  .name('No Console Logging')
  .description('Use Logger service instead of console')
  .severity('warning')
  .category('security')
  .tags('plugins', 'logging')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    const consoleMatches = ValidationHelpers.findMatches(
      context.content, 
      /console\.(log|warn|error|debug)/g
    );
    
    for (const match of consoleMatches) {
      violations.push(ValidationHelpers.createViolation(
        'no-console-logging',
        'warning',
        `Avoid console.${match.match.split('.')[1]}() - use Logger service instead`,
        context,
        { 
          suggestion: 'Import Logger from @nestjs/common and use proper logging',
          line: match.line,
          column: match.column
        }
      ));
    }
    
    return violations;
  });

export const noHardcodedSecrets = createRule()
  .id('no-hardcoded-secrets')
  .name('No Hardcoded Secrets')
  .description('No hardcoded secrets or credentials')
  .severity('error')
  .category('security')
  .tags('plugins', 'security', 'secrets')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    const secretPatterns = [
      /(?:api_key|apikey|password|secret|token)\s*[:=]\s*['"][^'"\n]{8,}['"]/g,
      /(?:sk_|pk_)[a-zA-Z0-9]{20,}/g,
      /[A-Za-z0-9+\/]{40,}/g // Base64 encoded secrets
    ];
    
    for (const pattern of secretPatterns) {
      const matches = ValidationHelpers.findMatches(context.content, pattern);
      for (const match of matches) {
        violations.push(ValidationHelpers.createViolation(
          'no-hardcoded-secrets',
          'error',
          'Potential hardcoded secret detected',
          context,
          { 
            suggestion: 'Use environment variables or configuration service for secrets',
            line: match.line,
            column: match.column
          }
        ));
      }
    }
    
    return violations;
  });

/**
 * Testing rules
 */
export const mustHaveTestFile = createRule()
  .id('must-have-test-file')
  .name('Must Have Test File')
  .description('Each plugin must have a corresponding test file')
  .severity('warning')
  .category('testing')
  .tags('plugins', 'testing')
  .filePattern(/src\/plugins\/.*\.plugin\.ts$/)
  .validate((context) => {
    const pluginContext = context as PluginValidationContext;
    const violations = [];
    const testFilePath = context.filePath.replace('.plugin.ts', '.plugin.spec.ts');
    
    try {
      require('fs').accessSync(testFilePath);
    } catch {
      violations.push(ValidationHelpers.createViolation(
        'must-have-test-file',
        'warning',
        `Missing test file: ${context.file.replace('.plugin.ts', '.plugin.spec.ts')}`,
        context,
        { 
          suggestion: 'Create corresponding test file for your plugin'
        }
      ));
    }
    
    return violations;
  });

/**
 * Export all NestJS plugin rules as a rule set
 */
export const nestjsPluginRules: ConstraintRule[] = [
  mustExtendBasePlugin,
  mustHaveInjectableDecorator,
  mustHaveMetadata,
  mustHaveSchema,
  mustHaveValidationRules,
  mustImplementExecuteMethod,
  mustImplementHealthCheck,
  mustImportRequiredModules,
  pluginFileNaming,
  pluginClassNaming,
  metadataCategoryValidation,
  noConsoleLogging,
  noHardcodedSecrets,
  mustHaveTestFile
];

File: nestjs-plugin-system/src/constraints/nestjs-plugin-rules.rb
#!/usr/bin/env ruby

require_relative '../../../constraint-experiment/lib/constraint_engine'

# NestJS Plugin Architecture Constraints
# Ensures LLMs generate plugins that follow our exact patterns

engine = ConstraintEngine.new

# ==========================================
# PLUGIN STRUCTURE CONSTRAINTS
# ==========================================

engine.rule "must_extend_base_plugin", "All plugins must extend BaseActionPlugin" do
  severity :error
  suggest "Extend BaseActionPlugin class: class YourPlugin extends BaseActionPlugin"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_class(/Plugin/) && !contains(/extends BaseActionPlugin/)
      violation! "Plugin class must extend BaseActionPlugin"
    end
  end
end

engine.rule "must_have_metadata", "Plugins must define readonly metadata property" do
  severity :error
  suggest "Add: readonly metadata: PluginMetadata = { name: '...', version: '...', ... }"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_class(/Plugin/) && !contains(/readonly metadata: PluginMetadata/)
      violation! "Plugin missing readonly metadata property"
    end
  end
end

engine.rule "must_have_schema", "Plugins must define readonly schema property" do
  severity :error
  suggest "Add: readonly schema: ActionSchema = { request: RequestType, response: ResponseType }"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_class(/Plugin/) && !contains(/readonly schema: ActionSchema/)
      violation! "Plugin missing readonly schema property"
    end
  end
end

engine.rule "must_have_validation_rules", "Plugins must define validation rules" do
  severity :error
  suggest "Add: readonly validationRules: ValidationRule[] = [...]"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_class(/Plugin/) && !contains(/readonly validationRules: ValidationRule/)
      violation! "Plugin missing readonly validationRules property"
    end
  end
end

engine.rule "must_implement_execute", "Plugins must implement execute method" do
  severity :error
  suggest "Add: async execute(input: any): Promise<any> { ... }"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_class(/Plugin/) && !has_method(/execute/)
      violation! "Plugin missing execute method"
    end
  end
end

engine.rule "must_implement_health_check", "Plugins must implement healthCheck method" do
  severity :error
  suggest "Add: async healthCheck(): Promise<boolean> { return true; }"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_class(/Plugin/) && !has_method(/healthCheck/)
      violation! "Plugin missing healthCheck method"
    end
  end
end

# ==========================================
# PLUGIN METADATA CONSTRAINTS
# ==========================================

engine.rule "metadata_must_be_complete", "Plugin metadata must include all required fields" do
  severity :error
  suggest "Include: name, version, description, author, category in metadata"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/readonly metadata: PluginMetadata/)
      required_fields = %w[name version description author category]
      required_fields.each do |field|
        unless contains(/#{field}:/)
          violation! "Plugin metadata missing required field: #{field}"
        end
      end
    end
  end
end

engine.rule "metadata_category_must_be_valid", "Plugin category must be valid" do
  severity :error
  suggest "Use one of: 'business', 'utility', 'integration', 'auth'"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/category:/)
      valid_categories = %w[business utility integration auth]
      category_line = find_pattern(/category:\s*['"]([^'"]+)['"]/).first
      if category_line
        category = category_line[:content].match(/category:\s*['"]([^'"]+)['"]/)[1]
        unless valid_categories.include?(category)
          violation! "Invalid category '#{category}'. Must be: #{valid_categories.join(', ')}"
        end
      end
    end
  end
end

engine.rule "metadata_name_must_match_class", "Plugin metadata name should match class name" do
  severity :warning
  suggest "Ensure metadata.name reflects the plugin class name"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    class_match = find_pattern(/class\s+(\w+Plugin)/).first
    name_match = find_pattern(/name:\s*['"]([^'"]+)['"]/).first
    
    if class_match && name_match
      class_name = class_match[:content].match(/class\s+(\w+Plugin)/)[1]
      metadata_name = name_match[:content].match(/name:\s*['"]([^'"]+)['"]/)[1]
      
      # Convert class name to kebab-case for comparison
      expected_name = class_name.gsub(/([A-Z])/, '-\1').downcase.sub(/^-/, '').sub(/-plugin$/, '')
      
      unless metadata_name.include?(expected_name) || expected_name.include?(metadata_name)
        violation! "Plugin name '#{metadata_name}' doesn't match class '#{class_name}'"
      end
    end
  end
end

# ==========================================
# SCHEMA CONSTRAINTS
# ==========================================

engine.rule "schema_must_have_types", "Plugin schema must define request and response types" do
  severity :error
  suggest "Define interfaces: interface RequestType {} and interface ResponseType {}"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/readonly schema: ActionSchema/)
      unless contains(/interface.*Request/) && contains(/interface.*Response/)
        violation! "Plugin must define Request and Response interface types"
      end
    end
  end
end

engine.rule "request_type_must_be_interface", "Request type must be a TypeScript interface" do
  severity :error
  suggest "Define: interface YourPluginRequest { ... }"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/request:/) && !contains(/interface.*Request/)
      violation! "Plugin request type must be defined as an interface"
    end
  end
end

engine.rule "response_type_must_be_interface", "Response type must be a TypeScript interface" do
  severity :error
  suggest "Define: interface YourPluginResponse { ... }"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/response:/) && !contains(/interface.*Response/)
      violation! "Plugin response type must be defined as an interface"
    end
  end
end

# ==========================================
# VALIDATION RULES CONSTRAINTS
# ==========================================

engine.rule "validation_rules_must_be_array", "Validation rules must be an array" do
  severity :error
  suggest "Define as: readonly validationRules: ValidationRule[] = [...]"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/validationRules:/) && !contains(/ValidationRule\[\]/)
      violation! "validationRules must be typed as ValidationRule[]"
    end
  end
end

engine.rule "validation_rules_must_have_structure", "Each validation rule must have name, message, and validate" do
  severity :error
  suggest "Use: { name: 'ruleName', message: 'Error message', validate: (input) => boolean }"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/validationRules:.*\[/)
      # Check for proper validation rule structure
      unless contains(/name:/) && contains(/message:/) && contains(/validate:/)
        violation! "Validation rules must have name, message, and validate properties"
      end
    end
  end
end

# ==========================================
# METHOD IMPLEMENTATION CONSTRAINTS
# ==========================================

engine.rule "execute_must_be_async", "Execute method must be async" do
  severity :error
  suggest "Define as: async execute(input: any): Promise<any>"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_method(/execute/) && !contains(/async execute/)
      violation! "execute method must be async"
    end
  end
end

engine.rule "execute_must_return_promise", "Execute method must return Promise" do
  severity :error
  suggest "Method signature: async execute(input: any): Promise<any>"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/execute.*:/) && !contains(/Promise</)
      violation! "execute method must return Promise"
    end
  end
end

engine.rule "health_check_must_be_async", "Health check must be async" do
  severity :error
  suggest "Define as: async healthCheck(): Promise<boolean>"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_method(/healthCheck/) && !contains(/async healthCheck/)
      violation! "healthCheck method must be async"
    end
  end
end

engine.rule "health_check_must_return_boolean", "Health check must return Promise<boolean>" do
  severity :error
  suggest "Method signature: async healthCheck(): Promise<boolean>"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/healthCheck.*:/) && !contains(/Promise<boolean>/)
      violation! "healthCheck method must return Promise<boolean>"
    end
  end
end

# ==========================================
# NESTJS DECORATOR CONSTRAINTS
# ==========================================

engine.rule "plugin_must_be_injectable", "Plugin class must have @Injectable decorator" do
  severity :error
  suggest "Add @Injectable() decorator above class declaration"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_class(/Plugin/) && !contains(/@Injectable/)
      violation! "Plugin class must have @Injectable() decorator"
    end
  end
end

engine.rule "must_import_injectable", "Must import Injectable from @nestjs/common" do
  severity :error
  suggest "Add: import { Injectable } from '@nestjs/common';"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/@Injectable/) && !contains(/import.*Injectable.*@nestjs\/common/)
      violation! "Must import Injectable from @nestjs/common"
    end
  end
end

engine.rule "must_import_base_plugin", "Must import BaseActionPlugin" do
  severity :error
  suggest "Add: import { BaseActionPlugin } from '../core/base-plugin.abstract';"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if contains(/extends BaseActionPlugin/) && !contains(/import.*BaseActionPlugin/)
      violation! "Must import BaseActionPlugin"
    end
  end
end

# ==========================================
# FILE STRUCTURE CONSTRAINTS
# ==========================================

engine.rule "plugin_file_naming", "Plugin files must follow naming convention" do
  severity :warning
  suggest "Name files as: feature-name.plugin.ts"
  
  when_file(/src\/plugins\//) do
    unless file_context.path.match?(/\.plugin\.ts$/)
      violation! "Plugin files should end with .plugin.ts"
    end
  end
end

engine.rule "plugin_class_naming", "Plugin classes must end with 'Plugin'" do
  severity :error
  suggest "Name classes as: FeatureNamePlugin"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    class_matches = find_pattern(/class\s+(\w+)/)
    class_matches.each do |match|
      class_name = match[:content].match(/class\s+(\w+)/)[1]
      unless class_name.end_with?('Plugin')
        violation! "Plugin class '#{class_name}' must end with 'Plugin'"
      end
    end
  end
end

engine.rule "one_plugin_per_file", "Each file should contain only one plugin class" do
  severity :warning
  suggest "Split multiple plugins into separate files"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    plugin_classes = find_pattern(/class\s+\w*Plugin/)
    if plugin_classes.size > 1
      violation! "File contains #{plugin_classes.size} plugin classes, should be one per file"
    end
  end
end

# ==========================================
# ERROR HANDLING CONSTRAINTS
# ==========================================

engine.rule "execute_must_handle_errors", "Execute method must have proper error handling" do
  severity :warning
  suggest "Wrap execution in try-catch block or use proper error boundaries"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    if has_method(/execute/) && !contains(/try.*catch|throw new Error/)
      violation! "execute method should include error handling"
    end
  end
end

engine.rule "no_console_logging", "Use proper logging instead of console" do
  severity :warning
  suggest "Use Logger service: import { Logger } from '@nestjs/common'"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    forbid("console logging").pattern(/console\.(log|warn|error|debug)/)
  end
end

# ==========================================
# SECURITY CONSTRAINTS
# ==========================================

engine.rule "no_eval_or_function_constructor", "No eval() or Function constructor" do
  severity :error
  suggest "Use safe alternatives for dynamic code execution"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    forbid("eval usage").pattern(/eval\s*\(/)
    forbid("Function constructor").pattern(/new Function\s*\(/)
  end
end

engine.rule "no_hardcoded_secrets", "No hardcoded secrets or credentials" do
  severity :error
  suggest "Use environment variables or configuration service"
  
  when_file(/src\/plugins\/.*\.ts$/) do
    forbid("API keys").pattern(/(?:api_key|apikey|password|secret|token)\s*[:=]\s*['"][^'"\n]{8,}['"]/)
  end
end

# ==========================================
# TESTING CONSTRAINTS
# ==========================================

engine.rule "plugin_must_have_test", "Each plugin must have a corresponding test file" do
  severity :warning
  suggest "Create test file: feature-name.plugin.spec.ts"
  
  when_file(/src\/plugins\/(.+)\.plugin\.ts$/) do
    plugin_name = file_context.path.match(/src\/plugins\/(.+)\.plugin\.ts$/)[1]
    test_file = "src/plugins/#{plugin_name}.plugin.spec.ts"
    
    unless File.exist?(test_file)
      violation! "Plugin missing test file: #{test_file}"
    end
  end
end

if __FILE__ == $0
  # Run NestJS plugin validation
  project_path = ARGV[0] || '.'
  
  puts "🔧 Running NestJS plugin architecture validation on #{project_path}..."
  violations = engine.check_project(project_path)
  
  if violations.empty?
    puts "✅ All plugins follow the architecture constraints!"
  else
    puts "⚠️  Found #{violations.size} plugin architecture violations:"
    engine.report_violations(violations)
    
    # Exit with error code if there are errors
    exit 1 if violations.any? { |v| v.severity == :error }
  end
end